File: DS200.Q21.1_Lab05/src/processing_server.py
"""
Processing Server - Object detection using Spark Streaming.

This server receives frames, performs person detection using YOLO,
and sends results to the storage server. Integrates with Apache Spark
Streaming for big data processing capabilities.
"""

import socket
import json
import logging
import threading
import base64
import io
import uuid
from datetime import datetime

# Spark imports
try:
    from pyspark import SparkContext
    from pyspark.streaming import StreamingContext
    SPARK_AVAILABLE = True
except ImportError:
    SPARK_AVAILABLE = False
    print("Warning: PySpark not available. Running in standalone mode.")

# OpenCV and detection imports
try:
    import cv2
    import numpy as np
    CV2_AVAILABLE = True
except ImportError:
    CV2_AVAILABLE = False
    print("Warning: OpenCV not available. Using mock detection.")

# YOLO imports
try:
    from ultralytics import YOLO
    YOLO_AVAILABLE = True
except ImportError:
    YOLO_AVAILABLE = False
    print("Warning: Ultralytics YOLO not available. Using mock detection.")

from config import Config, MessageType

# Configure logging
logging.basicConfig(level=Config.LOG_LEVEL, format=Config.LOG_FORMAT)
logger = logging.getLogger("ProcessingServer")


class PersonDetector:
    """Person detection using YOLO model."""
    
    def __init__(self, model_path=Config.YOLO_MODEL_PATH):
        self.model = None
        self.model_path = model_path
        self._load_model()
    
    def _load_model(self):
        """Load the YOLO model."""
        if YOLO_AVAILABLE:
            try:
                self.model = YOLO(self.model_path)
                logger.info(f"YOLO model loaded from {self.model_path}")
            except Exception as e:
                logger.warning(f"Could not load YOLO model: {e}. Using mock detection.")
                self.model = None
        else:
            logger.warning("YOLO not available. Using mock detection.")
    
    def detect(self, image_data):
        """
        Detect persons in the image.
        
        Args:
            image_data: Base64 encoded image or numpy array
            
        Returns:
            dict with person_count and bounding_boxes
        """
        if isinstance(image_data, str):
            # Decode base64 image
            try:
                image_bytes = base64.b64decode(image_data)
                nparr = np.frombuffer(image_bytes, np.uint8)
                image = cv2.imdecode(nparr, cv2.IMREAD_COLOR)
            except Exception as e:
                logger.error(f"Error decoding image: {e}")
                return self._mock_detection()
        else:
            image = image_data
        
        if self.model and CV2_AVAILABLE:
            return self._yolo_detection(image)
        else:
            return self._mock_detection()
    
    def _yolo_detection(self, image):
        """Perform actual YOLO detection."""
        try:
            results = self.model(image, classes=[0])  # class 0 is 'person'
            
            bounding_boxes = []
            for result in results:
                for box in result.boxes:
                    x, y, w, h = box.xywh[0].tolist()
                    confidence = float(box.conf[0])
                    
                    if confidence >= Config.CONFIDENCE_THRESHOLD:
                        bounding_boxes.append({
                            "x": int(x - w/2),
                            "y": int(y - h/2),
                            "width": int(w),
                            "height": int(h),
                            "confidence": round(confidence, 3)
                        })
            
            return {
                "person_count": len(bounding_boxes),
                "bounding_boxes": bounding_boxes
            }
        except Exception as e:
            logger.error(f"YOLO detection error: {e}")
            return self._mock_detection()
    
    def _mock_detection(self):
        """Generate mock detection results for testing."""
        import random
        num_persons = random.randint(0, 5)
        bounding_boxes = []
        
        for i in range(num_persons):
            bounding_boxes.append({
                "x": random.randint(50, 500),
                "y": random.randint(50, 400),
                "width": random.randint(40, 80),
                "height": random.randint(100, 180),
                "confidence": round(random.uniform(0.6, 0.99), 3)
            })
        
        return {
            "person_count": num_persons,
            "bounding_boxes": bounding_boxes
        }


class ProcessingServer:
    """TCP server for processing frames with Spark Streaming integration."""
    
    def __init__(self, host=Config.HOST, port=Config.PROCESSING_PORT):
        self.host = host
        self.port = port
        self.server_socket = None
        self.storage_connection = None
        self.running = False
        self.detector = PersonDetector()
        
        # Spark context
        self.sc = None
        self.ssc = None
    
    def start(self):
        """Start the processing server."""
        # Initialize Spark Streaming (optional)
        if SPARK_AVAILABLE:
            self._init_spark()
        
        # Start TCP server
        self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.server_socket.bind((self.host, self.port))
        self.server_socket.listen(5)
        self.running = True
        
        logger.info(f"Processing Server started on {self.host}:{self.port}")
        
        # Connect to storage server
        self._connect_to_storage_server()
        
        while self.running:
            try:
                client_socket, address = self.server_socket.accept()
                logger.info(f"Receiver connected from {address}")
                
                client_thread = threading.Thread(
                    target=self._handle_client,
                    args=(client_socket, address)
                )
                client_thread.daemon = True
                client_thread.start()
                
            except KeyboardInterrupt:
                logger.info("Shutting down processing server...")
                self.stop()
            except Exception as e:
                logger.error(f"Error accepting connection: {e}")
    
    def _init_spark(self):
        """Initialize Spark Streaming context."""
        try:
            self.sc = SparkContext(Config.SPARK_MASTER, Config.SPARK_APP_NAME)
            self.sc.setLogLevel("WARN")
            self.ssc = StreamingContext(self.sc, Config.SPARK_BATCH_INTERVAL)
            logger.info("Spark Streaming context initialized")
        except Exception as e:
            logger.warning(f"Could not initialize Spark: {e}")
            self.sc = None
            self.ssc = None
    
    def _connect_to_storage_server(self):
        """Establish connection to the storage server."""
        try:
            self.storage_connection = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.storage_connection.connect((Config.HOST, Config.STORAGE_PORT))
            logger.info(f"Connected to Storage Server at {Config.HOST}:{Config.STORAGE_PORT}")
        except ConnectionRefusedError:
            logger.warning("Storage server not available. Results will be logged only.")
            self.storage_connection = None
        except Exception as e:
            logger.error(f"Error connecting to storage server: {e}")
            self.storage_connection = None
    
    def _handle_client(self, client_socket, address):
        """Handle incoming frames from receiver server."""
        buffer = ""
        
        try:
            while self.running:
                data = client_socket.recv(Config.BUFFER_SIZE)
                if not data:
                    break
                    
                buffer += data.decode('utf-8')
                
                while '\n' in buffer:
                    message, buffer = buffer.split('\n', 1)
                    if message.strip():
                        self._process_message(message)
                        
        except Exception as e:
            logger.error(f"Error handling client {address}: {e}")
        finally:
            client_socket.close()
            logger.info(f"Client {address} disconnected")
    
    def _process_message(self, message):
        """Process incoming message and perform detection."""
        try:
            msg = json.loads(message)
            
            if msg.get('type') == MessageType.FRAME:
                frame_data = msg.get('data', {})
                self._process_frame(frame_data)
                
        except json.JSONDecodeError as e:
            logger.error(f"Invalid JSON: {e}")
        except Exception as e:
            logger.error(f"Error processing message: {e}")
    
    def _process_frame(self, frame_data):
        """Process a single frame for person detection."""
        frame_id = frame_data.get('frame_id', str(uuid.uuid4()))
        image_data = frame_data.get('data', '')
        
        logger.info(f"Processing frame: {frame_id}")
        
        # Perform detection
        start_time = datetime.now()
        detection_result = self.detector.detect(image_data)
        processing_time = (datetime.now() - start_time).total_seconds()
        
        # Build result
        result = {
            "frame_id": frame_id,
            "timestamp": datetime.now().isoformat(),
            "original_timestamp": frame_data.get('timestamp'),
            "person_count": detection_result['person_count'],
            "bounding_boxes": detection_result['bounding_boxes'],
            "processing_time_ms": round(processing_time * 1000, 2)
        }
        
        logger.info(f"Frame {frame_id}: Detected {result['person_count']} persons in {result['processing_time_ms']}ms")
        
        # Send to storage
        self._send_to_storage(result)
    
    def _send_to_storage(self, result):
        """Send detection result to storage server."""
        if self.storage_connection is None:
            self._connect_to_storage_server()
            
        if self.storage_connection:
            try:
                payload = {
                    "type": MessageType.DETECTION_RESULT,
                    "data": result
                }
                message = json.dumps(payload) + "\n"
                self.storage_connection.send(message.encode('utf-8'))
                logger.info(f"Sent result for frame {result['frame_id']} to storage")
            except BrokenPipeError:
                logger.error("Connection to storage server lost. Reconnecting...")
                self.storage_connection = None
            except Exception as e:
                logger.error(f"Error sending to storage: {e}")
        else:
            logger.warning(f"No storage connection. Result logged: {result}")
    
    def stop(self):
        """Stop the processing server."""
        self.running = False
        if self.ssc:
            self.ssc.stop(stopSparkContext=True, stopGraceFully=True)
        if self.storage_connection:
            self.storage_connection.close()
        if self.server_socket:
            self.server_socket.close()
        logger.info("Processing Server stopped")


def main():
    """Main entry point."""
    server = ProcessingServer()
    try:
        server.start()
    except KeyboardInterrupt:
        server.stop()


if __name__ == "__main__":
    main()

File: DS200.Q21.1_Lab05/spark/java/lab05-streaming/src/main/java/lab05/ProcessingServer.java
package lab05;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import org.apache.spark.SparkConf;
import org.apache.spark.streaming.Durations;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaReceiverInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.apache.spark.storage.StorageLevel;
import org.apache.spark.streaming.receiver.Receiver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.net.Socket;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.UUID;

/**
 * Processing Server with Spark Streaming integration.
 * 
 * Responsibilities:
 * - Receive frames from FrameReceiverServer via custom Receiver
 * - Process frames using Spark Streaming (distributed)
 * - Perform person detection (YOLO or mock)
 * - Send detection results to StorageServer
 * 
 * Big Data Context:
 * - Uses Spark Streaming for scalable stream processing
 * - Can distribute detection workload across cluster
 * - Processes frames in micro-batches for throughput
 */
public class ProcessingServer {
    private static final Logger logger = LoggerFactory.getLogger(ProcessingServer.class);
    private static final Gson gson = new Gson();
    
    private final int port;
    private final String storageHost;
    private final int storagePort;
    
    private JavaStreamingContext jssc;
    private Socket storageSocket;
    private PrintWriter storageWriter;
    
    public ProcessingServer() {
        this(Config.PROCESSING_PORT, Config.DEFAULT_HOST, Config.STORAGE_PORT);
    }
    
    public ProcessingServer(int port, String storageHost, int storagePort) {
        this.port = port;
        this.storageHost = storageHost;
        this.storagePort = storagePort;
    }
    
    /**
     * Start the processing server with Spark Streaming.
     */
    public void start() {
        logger.info("========================================");
        logger.info("Processing Server starting with Spark Streaming");
        logger.info("Listening on port: {}", port);
        logger.info("========================================");
        
        // Connect to storage server
        connectToStorageServer();
        
        // Initialize Spark Streaming
        SparkConf conf = new SparkConf()
                .setAppName(Config.SPARK_APP_NAME)
                .setMaster(Config.SPARK_MASTER)
                .set("spark.streaming.receiver.writeAheadLog.enable", "false")
                .set("spark.ui.enabled", "false");
        
        jssc = new JavaStreamingContext(conf, Durations.seconds(Config.SPARK_BATCH_INTERVAL));
        
        // Create custom receiver for TCP input
        JavaReceiverInputDStream<String> frameStream = jssc.receiverStream(
                new FrameReceiver(port));
        
        // Process each frame
        frameStream.foreachRDD(rdd -> {
            if (!rdd.isEmpty()) {
                rdd.collect().forEach(this::processFrame);
            }
        });
        
        // Start streaming
        jssc.start();
        logger.info("Spark Streaming started. Waiting for frames...");
        
        try {
            jssc.awaitTermination();
        } catch (InterruptedException e) {
            logger.info("Streaming interrupted");
        }
    }
    
    /**
     * Connect to StorageServer.
     */
    private void connectToStorageServer() {
        try {
            storageSocket = new Socket(storageHost, storagePort);
            storageWriter = new PrintWriter(
                new OutputStreamWriter(storageSocket.getOutputStream()), true);
            logger.info("Connected to Storage Server at {}:{}", storageHost, storagePort);
        } catch (IOException e) {
            logger.warn("Storage server not available. Results will be logged only.");
            storageSocket = null;
            storageWriter = null;
        }
    }
    
    /**
     * Process a single frame and perform detection.
     */
    private void processFrame(String jsonLine) {
        try {
            // Parse incoming message
            JsonObject msgObj = gson.fromJson(jsonLine, JsonObject.class);
            String msgType = msgObj.has("type") ? msgObj.get("type").getAsString() : "";
            
            if (!Config.MSG_TYPE_FRAME.equals(msgType)) {
                return;
            }
            
            JsonObject dataObj = msgObj.getAsJsonObject("data");
            DataModels.FrameData frame = gson.fromJson(dataObj, DataModels.FrameData.class);
            
            logger.info("Processing frame: {}", frame.frameId);
            
            long startTime = System.currentTimeMillis();
            
            // Perform detection (mock or real YOLO)
            List<DataModels.BoundingBox> boxes = detectPersons(frame);
            
            long processingTime = System.currentTimeMillis() - startTime;
            
            // Build result
            DataModels.DetectionResult result = new DataModels.DetectionResult();
            result.frameId = frame.frameId;
            result.timestamp = Instant.now().toString();
            result.originalTimestamp = frame.timestamp;
            result.personCount = boxes.size();
            result.boundingBoxes = boxes;
            result.processingTimeMs = processingTime;
            
            logger.info("Frame {}: Detected {} persons in {}ms", 
                    frame.frameId, result.personCount, processingTime);
            
            // Send to storage
            sendToStorage(result);
            
        } catch (Exception e) {
            logger.error("Error processing frame: {}", e.getMessage());
        }
    }
    
    /**
     * Detect persons in frame.
     * Currently uses mock detection. Replace with real YOLO for production.
     */
    private List<DataModels.BoundingBox> detectPersons(DataModels.FrameData frame) {
        // TODO: Integrate real YOLO detection here
        // For now, use mock detection for demonstration
        return mockDetection();
    }
    
    /**
     * Mock detection for testing without YOLO model.
     */
    private List<DataModels.BoundingBox> mockDetection() {
        List<DataModels.BoundingBox> boxes = new ArrayList<>();
        Random random = new Random();
        int numPersons = random.nextInt(6); // 0-5 persons
        
        for (int i = 0; i < numPersons; i++) {
            boxes.add(new DataModels.BoundingBox(
                    random.nextInt(500) + 50,   // x
                    random.nextInt(350) + 50,   // y
                    random.nextInt(40) + 40,    // width
                    random.nextInt(80) + 100,   // height
                    0.6 + random.nextDouble() * 0.39 // confidence 0.6-0.99
            ));
        }
        
        return boxes;
    }
    
    /**
     * Send detection result to StorageServer.
     */
    private synchronized void sendToStorage(DataModels.DetectionResult result) {
        if (storageWriter == null) {
            connectToStorageServer();
        }
        
        if (storageWriter != null) {
            try {
                DataModels.Message message = new DataModels.Message(Config.MSG_TYPE_RESULT, result);
                String json = gson.toJson(message);
                storageWriter.println(json);
                logger.info("Sent result for frame {} to StorageServer", result.frameId);
            } catch (Exception e) {
                logger.error("Error sending to storage: {}", e.getMessage());
                storageWriter = null;
            }
        } else {
            logger.warn("No storage connection. Result: {}", result);
        }
    }
    
    /**
     * Stop the processing server.
     */
    public void stop() {
        if (jssc != null) {
            jssc.stop(true, true);
        }
        try {
            if (storageSocket != null) storageSocket.close();
        } catch (IOException ignored) {}
        
        logger.info("Processing Server stopped");
    }
    
    /**
     * Custom Spark Receiver for TCP frame input.
     */
    public static class FrameReceiver extends Receiver<String> {
        private static final Logger logger = LoggerFactory.getLogger(FrameReceiver.class);
        private final int port;
        
        public FrameReceiver(int port) {
            super(StorageLevel.MEMORY_AND_DISK_2());
            this.port = port;
        }
        
        @Override
        public void onStart() {
            new Thread(this::receive).start();
        }
        
        @Override
        public void onStop() {
            // Cleanup handled by Spark
        }
        
        private void receive() {
            try (java.net.ServerSocket serverSocket = new java.net.ServerSocket(port)) {
                logger.info("FrameReceiver listening on port {}", port);
                
                while (!isStopped()) {
                    Socket socket = serverSocket.accept();
                    logger.info("FrameReceiver accepted connection from {}", 
                            socket.getRemoteSocketAddress());
                    
                    // Handle in separate thread
                    new Thread(() -> handleConnection(socket)).start();
                }
            } catch (Exception e) {
                if (!isStopped()) {
                    restart("Error in receiver: " + e.getMessage());
                }
            }
        }
        
        private void handleConnection(Socket socket) {
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(socket.getInputStream()))) {
                String line;
                while (!isStopped() && (line = reader.readLine()) != null) {
                    store(line);
                }
            } catch (IOException e) {
                logger.debug("Connection closed: {}", e.getMessage());
            }
        }
    }
    
    public static void main(String[] args) {
        ProcessingServer server = new ProcessingServer();
        
        Runtime.getRuntime().addShutdownHook(new Thread(server::stop));
        
        server.start();
    }
}
